### PR TITLE
report: render sub-rows in their own tr

### PR DESCRIPTION
**Pull request by connorjclark**
Part of #10369

Before this change, the sub rows columns didn't align well with each other, except in perfect conditions (short names, wide viewport). This PR makes it so each sub-row is rendered in its own `tr` element - the table layout does the alignment for us.

image

image

narrow:

image

### CODE REVIEW

**Comment by connorjclark** [comment]
image

weird flake bro

File: lighthouse-core/report/html/renderer/details-renderer.js [hunk 1]
```diff
@@ -375,49 +406,66 @@ class DetailsRenderer {
...
-          const valueElement =
-            value !== undefined && !Array.isArray(value) && this._renderTableValue(value, heading);
-          if (valueElement) valueFragment.appendChild(valueElement);
-        }
+        headingAndValuePairs.push({heading, value});
+      }
+
+      tbodyElem.append(this._renderRow(headingAndValuePairs));
+
+      // A single row data value can expand into multiple table rows. These additional table rows
```
Comment (paulirish) [inline lighthouse-core/report/html/renderer/details-renderer.js:423]:
```suggestion
      // A single details item can expand into multiple table rows. These additional table rows
```

File: lighthouse-core/report/html/renderer/details-renderer.js [hunk 2]
```diff
@@ -375,49 +406,66 @@ class DetailsRenderer {
 
     const tbodyElem = this._dom.createChildOf(tableElem, 'tbody');
     for (const row of details.items) {
-      const rowElem = this._dom.createChildOf(tbodyElem, 'tr');
+      // Render the main row.
```
Comment (paulirish) [inline lighthouse-core/report/html/renderer/details-renderer.js:409]: can you split out into a _renderTableRow() method?

File: lighthouse-core/report/html/renderer/details-renderer.js [hunk 3]
```diff
@@ -375,49 +406,66 @@ class DetailsRenderer {
...
-          };
-          const values = row[subRowsHeading.key];
-          if (Array.isArray(values)) {
-            const subRowsElement = this._renderSubRows(values, subRowsHeading);
-            valueFragment.appendChild(subRowsElement);
+      if (!subRowHeadings.some(Boolean)) continue;
+
+      // All sub-row data arrays should be the same length, but just in case they are not,
+      // determine the longest data array and fill the missing values with an
+      // null pair (which creates an empty cell).
```
Comment (paulirish) [inline lighthouse-core/report/html/renderer/details-renderer.js:441]: adding nulls to create empty cells seems very weird to me. 
its to maintain alignment with the parent headings right? but dont the parent headings already know what their subheadings are? can we use that relationship?

File: lighthouse-core/report/html/report-styles.css [hunk 4]
```diff
@@ -1425,12 +1425,12 @@
   text-decoration: underline dotted #999;
 }
 
-.lh-sub-rows:not(:first-child) .lh-sub-row {
```
Comment (paulirish) [inline lighthouse-core/report/html/report-styles.css:1428]: i kinda miss the previous styling.
i wonder if there's a way we can still have zebra striping on the non-subrows and the subrows feel like they belong.

... trying a few things out it seems impossible without way too much time. ah well.

File: lighthouse-core/report/html/report-styles.css [hunk 5]
```diff
@@ -1425,12 +1425,12 @@
...
+.lh-sub-row td:first-child {
+  display: block;
   margin-left: 20px;
-  color: var(--color-gray-700);
 }
-.lh-sub-row {
-  margin-bottom: 0;
+.lh-sub-row td {
+  color: var(--color-gray-700);
```
Comment (paulirish) [inline lighthouse-core/report/html/report-styles.css:1433]: drop font-size down 1 or 2 px?

File: lighthouse-core/report/html/report-styles.css (hunk 4 above)
Comment (connorjclark) [reply lighthouse-core/report/html/report-styles.css:1428]: > i kinda miss the previous styling.

I also prefer the previous styling. It's simple if the rendering code adds an "even" or "odd" class to all rows generated from a single data value. Then we just need to duplicate the CSS for reach even / odd.

Comment (paulirish) [reply lighthouse-core/report/html/report-styles.css:1428]: > It's simple if the rendering code adds an "even" or "odd" class to all rows generated from a single data value.

true!  i don't mind that.

File: lighthouse-core/report/html/renderer/details-renderer.js [hunk 6]
```diff
@@ -375,49 +406,66 @@ class DetailsRenderer {
...
-            valueType: heading.subRows.valueType || heading.valueType,
-            granularity: heading.subRows.granularity || heading.granularity,
-            displayUnit: heading.subRows.displayUnit || heading.displayUnit,
-            label: '',
-          };
-          const values = row[subRowsHeading.key];
-          if (Array.isArray(values)) {
-            const subRowsElement = this._renderSubRows(values, subRowsHeading);
-            valueFragment.appendChild(subRowsElement);
+      if (!subRowHeadings.some(Boolean)) continue;
```
[the remark this answers is not in the document]
Comment (connorjclark) [reply lighthouse-core/report/html/renderer/details-renderer.js:437]: there is a `return null;` in the map

File: lighthouse-core/report/html/renderer/details-renderer.js (hunk 3 above)
Comment (connorjclark) [reply lighthouse-core/report/html/renderer/details-renderer.js:441]: background color only colors the width of the table if there is an element there.

File: lighthouse-core/report/html/report-styles.css (hunk 4 above)
Comment (connorjclark) [reply lighthouse-core/report/html/report-styles.css:1428]: done

File: lighthouse-core/report/html/renderer/details-renderer.js (hunk 3 above)
Comment (connorjclark) [reply lighthouse-core/report/html/renderer/details-renderer.js]: Paul refactor to remove the nulls here: https://github.com/GoogleChrome/lighthouse/compare/subrow-layout...subrow-layout-by-col

I think most of the complexity in your and my impl. comes from trying to pass the values to this `_renderRow` function. How about instead the fn uses a callback? I think it'd look like this:

```
_renderRow(headings, heading => row && row[heading.key])

for i in numSubRow:
  const subRowEl = _renderRow(headings, heading => row && Array.isArray(row[heading.key]) && row[heading.key][i])
  ...
```

Comment (paulirish) [reply lighthouse-core/report/html/renderer/details-renderer.js]: yeah thats a nice solution. :)

Comment (brendankenny) [reply lighthouse-core/report/html/renderer/details-renderer.js]: Peanut gallery: this seems like it's getting pretty involved for a static, self-generated table json of known width and height being turned into a `<table>` of known width and height. There's some table cruft we have to live with for legacy support, but subrows are brand new :)

Since (I don't believe) subrows are turned on anywhere yet (and we haven't shipped 6.0 yet regardless), maybe knowing more of the use case now gives a chance to step back and massage the data structure a bit to make dealing with them simpler? (#&#x2060;5 on the classic https://users.ece.utexas.edu/~adnan/pike.html)

**Comment by paulirish** [comment]
Awaiting brendan's typecheck magic for subRows. Will enable a new details subrow shape.

**Comment by paulirish** [comment]
This was the draft data shape for subrows we discussed last time.

image

one `subRows` array (instead of multiple arrays).

**Comment by connorjclark** [comment]
might also go with `subItems: [...]`

**Comment by connorjclark** [comment]
@brendankenny do you have an update on the types? If you can't devote any time in the next couple weeks can you possibly share what you've done so far so that I could prioritize it?

**Comment by brendankenny** [comment]
> @brendankenny do you have an update on the types? If you can't devote any time in the next couple weeks can you possibly share what you've done so far so that I could prioritize it?

I was hoping to iterate on this more, but lots of things have gotten in the way :) Maybe this is good enough or you can find some ways to improve.

tl;dr: branch with audits and some of the tests converted to try out the proposal: https://github.com/GoogleChrome/lighthouse/compare/master...subrowstraw

The biggest constraint is what we identified in the beginning: since `TableItem` has a generic index (so table columns can be named whatever they want to be), all item properties have to have the union of all possible item types as at least their base type (though individual properties can be constrained within that base type).

That means we can't have a unique `subRows` property with a unique form (if it has that form, every other `TableItem` property also has to be able to take that form). IIRC, @connorjclark this is what you originally considered and the downside is why you didn't pursue that idea.

The other constraint is for downstream renderers: ideally if a report renderer doesn't support subrows yet, they'll just silently be ignored.

So, possible approaches:
1. Rather than having a `TableItem = {[p: string]: undefined | ItemValue}` indexed type, enumerate all possible table item properties. `TableItem = {subRows?: TableSubRow[], url?: string, transferSize?: number, location?: SourceLocationValue, ...}`. I tried this to see what it would look like and we already have an enormous number of these and would require either reusing existing item property names or adding to `audit-details.d.ts` every time a new audit table is added. Doable but fairly ridiculous? We really do want it to be an indexed type.

1. Have a specific `TableItems['subRows']` property with a`TableSubRow[]` type, but have to let any other `TableItem` property take a `TableSubRow[]` type, even if they're never that. Feels bad type-wise but can enforce that it's only a `subRows` property at runtime like we do with a few other aspects of details rendering.

1. Let only a `TableItems['subRows']` take a `TableSubRow[]`, `@ts-ignore` all the places where this breaks the type system (let's not do this one :)

1. Current solution on master with downsides discussed above

1. Other solutions outside of `items`. e.g. maybe a parallel `subItems`, but would require bookkeeping of indices between which `subItems` are connected to which items.

After checking out (1) and not liking it, I tried out (2) and I'm reasonably happy. With this proposal, subrows become a bit more predictable (looking and acting exactly like `items`), easier to iterate on in the report renderer, and still have the desired property that renderers that don't support subrows yet will just ignore them by default. Downside is looseness of `TableSubRows` showing up in places we don't want it, but I don't think that's a big issue in practice.

Heart of things:
```ts
/** A table item for rows that are nested within a top-level TableItem (row). */
export interface TableSubRows {
  type: 'subrows';
  items: TableItem[];
}

/** Possible types of values found within table items. */
type ItemValue = string | number | boolean | DebugData | NodeValue | SourceLocationValue | LinkValue | UrlValue | CodeValue | TableSubRows;

export interface TableItem {
  debugData?: DebugData;
  subRows?: TableSubRows;
  [p: string]: undefined | ItemValue;
}
```

and an item with subrows would look something like
```json
{
  "url": "http://example.com/example.js",
  "totalBytes": 71654,
  "wastedBytes": 30470,
  "wastedPercent": 42.52388078488413,
  "subRows": {
    "type": "subrows",
    "items": [{
        "sources": "...Control/assets/js/vendor/katex.min.js",
        "sourceBytes": 14773,
        "sourceWastedBytes": 500,
    }, {
      "//": "..."
    }],
  },
}
```
